### ISSUE TITLE
Difference between `SentenceLabelDataset` and `GroupByLabelBatchSampler`

### ISSUE THREAD

**Issue by vibhas-singh** [UKPLab/sentence-transformers#2920]
Hi @tomaarsen 
First of all - kudos to you maintaining such a pragmatic library.

I am facing some difficulty on using `GROUP_BY_LABEL` batch sampler in v3.0 and want to highlight the issues to check if there is any way to mitigate those.

I went through the issues and found this: https://github.com/UKPLab/sentence-transformers/issues/2698#issuecomment-2144534374
You have mentioned here that the idea is to replace `SentenceLabelDataset` by `GroupByLabelBatchSampler` but I think there is very drastic differences between the two and we haven't retained the same functionality of `SentenceLabelDataset` while opting for `GroupByLabelBatchSampler` as a replacement.

I am taking an detailed example to explain the differences:

Let's take a simple example with a list of integers representing classes, and we'll use it to illustrate how the two approaches handle homogeneity in batch construction.

### Example Data:
- **Classes:** `[1, 1, 1, 1, 2, 2, 2, 3, 3, 4]`
- **Number of Samples per Label:**
  - Class 1: 4 samples
  - Class 2: 3 samples
  - Class 3: 2 samples
  - Class 4: 1 sample
- **Batch Size:** 4
- **Samples per Label in the Second Method:** 2

#### **Step 1: Initialization**
- **Grouping by Labels:**
  - Class 1: `[0, 1, 2, 3]`
  - Class 2: `[4, 5, 6]`
  - Class 3: `[7, 8]`
  - Class 4: `[9]`

- **Truncate to Even Number:**
  - Class 1: `[0, 1, 2, 3]` (all 4 samples)
  - Class 2: `[4, 5]` (only 2 samples, 1 sample is discarded)
  - Class 3: `[7, 8]` (both samples)
  - Class 4: `[9]` (removed because it doesn't have at least 2 samples)

- **Resulting Groups:**
  - Class 1: `[0, 1, 2, 3]`
  - Class 2: `[4, 5]`
  - Class 3: `[7, 8]`

#### **Step 2: Batch Construction**
- **Batch 1:**
  - Randomly selects Class 1 and takes all samples: `[0, 1, 2, 3]`
  - **Result:** `[0, 1, 2, 3]` (4 samples from Class 1, fully homogeneous)

- **Batch 2:**
  - Randomly selects Class 2 and takes both samples: `[4, 5]`
  - Randomly selects Class 3 and takes both samples: `[7, 8]`
  - **Result:** `[4, 5, 7, 8]` (2 samples from Class 2, 2 samples from Class 3)

#### **Step 1: Initialization**
- **Label Range:** `[1, 2, 3, 4]`
- **Already Seen Dictionary:** Empty

#### **Step 2: Batch Construction**
- **Batch 1:**
  - Starts with Class 1, takes 2 samples (e.g., `[0, 1]`)
  - **Result:** `[0, 1]` (2 samples from Class 1)
  - Moves to Class 2, takes 2 samples (e.g., `[4, 5]`)
  - **Result:** `[0, 1, 4, 5]` (2 samples from Class 1, 2 samples from Class 2)

- **Batch 2:**
  - Continues with Class 1, takes remaining 2 samples (e.g., `[2, 3]`)
  - **Result:** `[2, 3]` (2 samples from Class 1)
  - Continues with Class 2 but only 1 sample is left, so skips it and moves to Class 3
  - Takes both samples from Class 3 (e.g., `[7, 8]`)
  - **Result:** `[2, 3, 7, 8]` (2 samples from Class 1, 2 samples from Class 3)

- **Batch 3:**
  - If continuing, it would reset, shuffle, and start over, potentially including Class 4 depending on whether `with_replacement` is `True` or `False`.

### **Comparison of Homogeneity:**

- **`GroupByLabelBatchSampler`:**
  - **Batch 1:** Fully homogeneous batch from Class 1: `[0, 1, 2, 3]`.
  - **Batch 2:** Mixes samples from two classes (Class 2 and Class 3) because neither can fully fill a batch on their own: `[4, 5, 7, 8]`.
  - **Overall:** Prioritizes keeping batches homogeneous when possible, especially when a class has enough samples to fill an entire batch. It avoids mixing classes unless necessary.

- **`SentenceLabelDataset`:**
  - **Batch 1:** Mixed from the start, taking 2 samples from Class 1 and 2 from Class 2: `[0, 1, 4, 5]`.
  - **Batch 2:** Also mixed, with samples from Class 1 and Class 3: `[2, 3, 7, 8]`.
  - **Overall:** More likely to mix classes in every batch. It enforces a fixed number of samples per label and sequentially processes labels, leading to less homogeneous batches overall.

# TL;DR:
I am trying to fine-tune sentence transformers models using the dataset with this label distribution:
```
Class 1: 5000 Samples
Class 2: 5000 Samples
Class 3: 3000 Samples
Class 5 to 50: Less than 50 samples each
```
In the new `GroupByLabelBatchSampler` the batching logic is yielding most of the batches as homogeneous and there is not much improvement observed after fine-tuning. 
IMO this type of data could have been easily used with `SentenceLabelDataset` as it ensures there is at max `N` samples from each label in a batch. Intuitively, ST models should benefit from having in-batch negatives and more heterogeneous batches.

Can you help me in veryfying if my understanding is correct and if yes, is there any way to opt for the older logic?

**Comment by tomaarsen** [comment]
Hello!

You're very right in your analysis: `GroupByLabelBatchSampler` was designed to replace `SentenceLabelDataset`, and the former is homogeneous whereas the latter is not so much. For reference, here is the docstring for the new `GroupByLabelBatchSampler`: https://github.com/UKPLab/sentence-transformers/blob/0a32ec8445ef46b2b5d4f81af4931e293d42623f/sentence_transformers/sampler.py#L40-L43

This sampler is meant for the `Batch...TripletLoss` classes, which require that each batch contains at least 2 examples per label class. These losses compare across all samples with the same label within the same batch, benefiting from 1) larger batches and 2) more samples with the same label in the each batch. At least, that is my understanding.
As a result, in theory a more homogeneous batch should result in a better training signal for these losses. However, I admit that I haven't tested it out in practice, and I may be wrong.

> Is there any way to opt for the older logic?

Yes, and no. You can override the Trainer's `get_batch_sampler`: https://github.com/UKPLab/sentence-transformers/blob/0a32ec8445ef46b2b5d4f81af4931e293d42623f/sentence_transformers/trainer.py#L459-L466

And replace it with a function that immediately returns a custom Batch Sampler which has your desired behaviour. So yes: you can use the older logic, but no: you'd have to write it yourself.

Hope this helps a bit.

- Tom Aarsen

**Comment by MrLoh** [comment]
I ran into the same issue and dug into it a bit deeper. Turns out the homogeneous batching doesn't just give a weaker training signal, it gives **no signal at all** for the `Batch*TripletLoss` classes that `GroupByLabelBatchSampler` is recommended for.

## What's happening

As @vibhas-singh showed, `GroupByLabelBatchSampler` creates mostly single-class batches. The [`__iter__`](https://github.com/huggingface/sentence-transformers/blob/master/sentence_transformers/sampler.py#L157-L153) just dumps all samples for each label into a buffer and chunks it:

```python
for label_idx in torch.randperm(len(self.groups), generator=self.generator):
    label = unique_labels[label_idx]
    samples = self.groups[label]
    partial_batch.extend(samples)            # all samples for this label at once
    while len(partial_batch) >= self.batch_size:
        yield partial_batch[: self.batch_size]  # mostly single-class
        partial_batch = partial_batch[self.batch_size :]
```

You only get mixed-class batches at label boundaries — when leftover samples from one label get combined with the next. For any class with `count >= batch_size`, that means `count // batch_size` pure single-class batches before you maybe get one mixed batch.

The problem is that `BatchHardSoftMarginTripletLoss` (and siblings) need at least 2 distinct classes per batch to mine valid triplets. With a single-class batch there are no negatives, so you get:

- loss = ln(2) ≈ 0.6931 (degenerate soft-margin value)
- grad_norm = 0.0

@tomaarsen mentioned [above](https://github.com/UKPLab/sentence-transformers/issues/2920#issuecomment-2326919693):

> in theory a more homogeneous batch should result in a better training signal for these losses. However, I admit that I haven't tested it out in practice

So I tested it — the model simply doesn't learn.

## Numbers

I used the [`osanseviero/twitter-airline-sentiment`](https://huggingface.co/datasets/osanseviero/twitter-airline-sentiment) dataset (13k samples, 3 classes: 8260 negative / 2789 neutral / 2127 positive). Full reproduction script at the bottom.

### Batch composition: 99–100% single-class

```
batch_size=32:  410/412 single-class batches (100%)
batch_size=64:  204/206 single-class batches (99%)
batch_size=128: 101/103 single-class batches (98%)
```

Only 2 batches per epoch (at label transitions) contain multiple classes, regardless of batch size.

### Training: GROUP_BY_LABEL vs default sampler

Training `distilbert-base-uncased` with `BatchHardSoftMarginTripletLoss`, `batch_size=64`, 1 epoch:

**GROUP_BY_LABEL** — 204/206 micro-batches are single-class, **every logged step has zero gradient**:

```
step  10 | loss=0.6931 | grad_norm=0.0000  <-- ZERO GRADIENT
step  20 | loss=0.6931 | grad_norm=0.0000  <-- ZERO GRADIENT
step  30 | loss=0.6931 | grad_norm=0.0000  <-- ZERO GRADIENT
step  40 | loss=0.6931 | grad_norm=0.0000  <-- ZERO GRADIENT
step  50 | loss=0.8634 | grad_norm=0.0000  <-- ZERO GRADIENT
  ...all 20 logged steps are zero...
step 200 | loss=0.6931 | grad_norm=0.0000  <-- ZERO GRADIENT
```

**Default sampler (random)** — 0/206 micro-batches are single-class, loss decreases immediately:

```
step  10 | loss=1.4111 | grad_norm=3.4437
step  20 | loss=0.9080 | grad_norm=1.1400
step  30 | loss=0.7765 | grad_norm=1.4319
step  40 | loss=0.7463 | grad_norm=0.4985
  ...loss keeps decreasing...
step 200 | loss=0.7156 | grad_norm=0.2011
```

## The docstring contradiction

The [docstring](https://github.com/huggingface/sentence-transformers/blob/master/sentence_transformers/sampler.py#L69-L93) says two things that contradict each other:

1. "aims to create batches such that each batch contains samples where the labels are **as homogeneous as possible**"
2. "meant to be used alongside the `Batch...TripletLoss` classes, which require that each batch contains **at least 2 examples per label class**"

The first one is what it actually does. The second is a requirement it doesn't satisfy.

## I'm planning a PR

I'd like to add a `BalancedLabelBatchSampler` using PK sampling (P labels × K samples per label per batch), keeping `GroupByLabelBatchSampler` as-is for backward compatibility. The idea:

- `P = min(num_labels, batch_size // 2)` — as many labels as fit with K >= 2
- `K = batch_size // P` — samples per label
- Every batch guaranteed to have multiple classes with 2+ examples each
- New `BatchSamplers.BALANCED_LABELS` enum value, wired through `trainer.py`
- Docs updated to recommend `BALANCED_LABELS` for `Batch*TripletLoss`

This is basically what `SentenceLabelDataset` used to do before it was deprecated.

<details>
<summary>Full reproduction script</summary>

```python
"""
Reproduction: GroupByLabelBatchSampler produces single-class batches,
causing zero gradients with Batch*TripletLoss.

Uses osanseviero/twitter-airline-sentiment (13k samples, 3 classes).

Run: python group_by_label_issue.py
Requires: sentence-transformers>=3.0, datasets
"""

from collections import Counter

from datasets import ClassLabel, Dataset, load_dataset
from sentence_transformers import SentenceTransformer, SentenceTransformerTrainer, losses
from sentence_transformers.sampler import GroupByLabelBatchSampler
from sentence_transformers.training_args import BatchSamplers, SentenceTransformerTrainingArguments
from transformers.trainer_callback import TrainerCallback


def load_airline_sentiment():
    ds = load_dataset("osanseviero/twitter-airline-sentiment", split="train")
    ds = ds.rename_columns({"text": "sentence", "airline_sentiment": "label"})
    ds = ds.cast_column("label", ClassLabel(names=sorted(set(ds["label"]))))
    ds = ds.select_columns(["sentence", "label"])
    split = ds.train_test_split(test_size=0.1, seed=42, stratify_by_column="label")
    return split["train"], split["test"]


class BatchLabelLogger(TrainerCallback):
    def __init__(self):
        self.micro_batch_labels: list[Counter] = []
        self.step_logs: list[dict] = []


class LabelTrackingTrainer(SentenceTransformerTrainer):
    def __init__(self, *args, label_logger: BatchLabelLogger, **kwargs):
        super().__init__(*args, **kwargs)
        self.label_logger = label_logger

    def training_step(self, model, inputs, num_items_in_batch=None):
        if "label" in inputs:
            self.label_logger.micro_batch_labels.append(
                Counter(inputs["label"].tolist())
            )
        return super().training_step(model, inputs, num_items_in_batch)


def train_and_report(train_ds, test_ds, sampler_name, batch_sampler):
    model = SentenceTransformer("distilbert-base-uncased")
    label_logger = BatchLabelLogger()

    args = SentenceTransformerTrainingArguments(
        output_dir=f"/tmp/triplet_debug_{sampler_name}",
        per_device_train_batch_size=64,
        per_device_eval_batch_size=64,
        num_train_epochs=1,
        logging_steps=10,
        batch_sampler=batch_sampler,
        save_strategy="no",
        eval_strategy="no",
        report_to="none",
    )

    class LogCapture(TrainerCallback):
        def on_log(self, args, state, control, logs=None, **kwargs):
            if logs and "loss" in logs and "eval_loss" not in logs:
                label_logger.step_logs.append({
                    "step": state.global_step,
                    "loss": logs.get("loss"),
                    "grad_norm": logs.get("grad_norm"),
                    "epoch": logs.get("epoch"),
                })

    trainer = LabelTrackingTrainer(
        model=model,
        loss=losses.BatchHardSoftMarginTripletLoss(model),
        args=args,
        train_dataset=train_ds,
        eval_dataset=test_ds,
        callbacks=[LogCapture()],
        label_logger=label_logger,
    )
    trainer.train()
    return label_logger


def main():
    print("Loading twitter airline sentiment dataset...\n")
    train_ds, test_ds = load_airline_sentiment()

    # Part 1: Batch composition
    print("=" * 70)
    print("Part 1: Batch composition with GroupByLabelBatchSampler")
    print("=" * 70)
    labels = train_ds["label"]
    label_counts = Counter(labels)
    label_names = train_ds.features["label"].names
    print(f"Dataset: {len(train_ds)} samples, {len(label_counts)} classes")
    for label_id, count in sorted(label_counts.items()):
        print(f"  {label_names[label_id]}: {count}")
    for batch_size in [32, 64, 128]:
        sampler = GroupByLabelBatchSampler(
            dataset=train_ds, batch_size=batch_size, drop_last=False, valid_label_columns=["label"]
        )
        total = single_class = 0
        for batch_indices in sampler:
            total += 1
            if len(set(labels[i] for i in batch_indices)) == 1:
                single_class += 1
        print(f"batch_size={batch_size}: {single_class}/{total} single-class ({single_class/total:.0%})")
    print()

    # Part 2: Training comparison
    print("=" * 70)
    print("Part 2: Training BatchHardSoftMarginTripletLoss")
    print("=" * 70)
    print(f"Train: {len(train_ds)}, Test: {len(test_ds)}, batch_size=64, 1 epoch\n")
    for name, sampler in [
        ("GROUP_BY_LABEL", BatchSamplers.GROUP_BY_LABEL),
        ("Default (random)", BatchSamplers.BATCH_SAMPLER),
    ]:
        print(f"--- {name} ---")
        logger = train_and_report(train_ds, test_ds, name, sampler)
        single = sum(1 for c in logger.micro_batch_labels if len(c) == 1)
        multi = sum(1 for c in logger.micro_batch_labels if len(c) > 1)
        zero = sum(1 for log in logger.step_logs if log["grad_norm"] == 0.0)
        print(f"Micro-batches: {single} single-class, {multi} multi-class")
        print(f"Steps with zero gradient: {zero}/{len(logger.step_logs)}")
        for log in logger.step_logs[:20]:
            marker = " <-- ZERO GRADIENT" if log["grad_norm"] == 0.0 else ""
            print(f"  step {log['step']:3d} | loss={log['loss']:.4f} | grad_norm={log['grad_norm']:.4f}{marker}")
        if len(logger.step_logs) > 20:
            print(f"  ... ({len(logger.step_logs) - 20} more steps)")
        print()


if __name__ == "__main__":
    main()
```

</details>

**Environment:** sentence-transformers 5.1.0, torch 2.6.0, Python 3.12

**Comment by MrLoh** [comment]
I went ahead and put together a #3668 that adds a `BalancedLabelBatchSampler` using PK sampling. @tomaarsen would appreciate your eyes on this when you get a chance! Happy to adjust the approach if you'd prefer a different direction (e.g. fixing `GroupByLabelBatchSampler` directly instead of adding a new sampler).
